File: Week001/acruikshank/match.hs
prefix :: String -> String -> Maybe String
prefix s [] = Just s
prefix [] p = Nothing
prefix (s:ss) (p:ps)
  | s == p = prefix ss ps
  | otherwise = Nothing

match :: String -> String -> Int -> Bool
match [] pattern n = n == 0
match (s:ss) pattern n = case (prefix (s:ss) pattern) of
  Nothing -> match ss pattern n
  Just rest -> match rest pattern (n-1)

main = do
  if not (match "abcabc" "abc" 1)       then putStrLn "PASS" else putStrLn "FAIL"
  if match "abcabc" "abc" 2             then putStrLn "PASS" else putStrLn "FAIL"
  if match "Hello Jello" "ello" 2       then putStrLn "PASS" else putStrLn "FAIL"
  if not (match "Hello Jello" "ello" 3) then putStrLn "PASS" else putStrLn "FAIL"
  if not (match "Ratatattat" "at" 3)    then putStrLn "PASS" else putStrLn "FAIL"
  if match "Ratatattat" "at" 4          then putStrLn "PASS" else putStrLn "FAIL"
  if not (match "oooo" "ooo" 2)         then putStrLn "PASS" else putStrLn "FAIL"
  if not (match "faaaaat" "fat" 1)      then putStrLn "PASS" else putStrLn "FAIL"
  if (match "eeep" "eep" 1)             then putStrLn "PASS" else putStrLn "FAIL"

File: Week001/acruikshank/match.js

function match(string, pattern, count) {
  return !string.length ? count==0 :
    string.slice(0, pattern.length) === pattern ?
      match( string.slice(pattern.length), pattern, count-1 ) :
      match( string.slice(1), pattern, count )
}

function test(value) { console.log( value ? "PASS" : "FAIL" ); }

test( ! match('abcabc', 'abc', 1) );
test( match('abcabc', 'abc', 2) );
test( match('Hello Jello', 'ello', 2) );
test( ! match('Hello Jello', 'ello', 3) );
test( ! match('Ratatattat', 'at', 3) );
test( match('Ratatattat', 'at', 4) );
test( ! match('oooo', 'ooo', 2) );